File: API/PetCenterModels/DataTransferObjects/BreedDTO.cs
using Microsoft.VisualBasic;
using PetCenterModels.DBTables;
using PetCenterModels.DataTransferObjects;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using PetCenterModels.SearchObjects;

namespace PetCenterModels.DataTransferObjects
{
    public class BreedDTO : ISerializableRequestDTO<Breed>, IAlbumCarryingDTO<Breed,BreedDTO>
    {
       
        public Guid? Id {get; set;} = null;
        
        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public Guid KindId {get; set;} = Guid.Empty;

        public AnimalScale Scale {get; set;}

        public float Investment {get; set;} = 0.0f;
        public float Territory {get; set;} = 0.0f;
        public float Pricing {get; set;} = 0.0f;
        public float Longevity {get; set;} = 0.0f;
        public float Cohabitation {get; set;} = 0.0f;

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public string Title {get; set;} = string.Empty;

        public Guid AlbumId {get; set;} = Guid.Empty;

        public List<ImageDTO> Images {get; set;} = new();


        public static BreedDTO? FromEntity(Breed? entity)
        {
            if(entity==null){return null;}
            BreedDTO output = new BreedDTO
            {
                Id = entity.Id,
                CurrentVersion=entity.CurrentVersion,
                Title = entity.Title,
                KindId=entity.KindId,
                AlbumId=entity.AlbumId,
                Investment=entity.Investment,
                Territory=entity.Territory,
                Pricing=entity.Pricing,
                Longevity=entity.Longevity,
                Cohabitation=entity.Cohabitation,
                Scale=entity.Scale
            };

            if (entity.Album != null)
            {
                output.Images = entity.Album.Images.Select(i=>ImageDTO.FromEntity(i)!).ToList();
            }

            return output;
        }

        public Breed? ToEntity()
        {
            Breed breed = new();
            breed.CurrentVersion=CurrentVersion;
            breed.KindId=KindId;
            breed.AlbumId=AlbumId;
            breed.Scale=Scale;
            breed.Investment=Investment;
            breed.Territory=Territory;
            breed.Cohabitation=Cohabitation;
            breed.Longevity=Longevity;
            breed.Pricing=Pricing;
            breed.Title=Title;
            return breed;
        }
        
        
        public bool Validate()
        {
            Investment = Math.Clamp(Investment,0.0f,1.0f);
            Territory = Math.Clamp(Territory,0.0f,1.0f);
            Pricing = Math.Clamp(Pricing,0.0f,1.0f);
            Longevity = Math.Clamp(Longevity,0.0f,1.0f);
            Cohabitation = Math.Clamp(Cohabitation,0.0f,1.0f);
            return !string.IsNullOrWhiteSpace(Title);
        }



    }
}

File: Frontend/pet_center_app/lib/models/data_transfer/breed_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pet_center_app/models/data_transfer/note_sub_dto.dart';
import 'package:pet_center_app/models/data_transfer/image_dto.dart';
import 'package:pet_center_app/models/enums.dart';
part 'breed_dto.g.dart';

@JsonSerializable(explicitToJson: true)
class BreedDTO {
  String? id;
  String currentVersion;
  String kindId;
  AnimalScale scale;
  double investment;
  double territory;
  double pricing;
  double longevity;
  double cohabitation;
  List<NoteSubDTO>? notes;
  String title;
  String albumId;
  List<ImageDTO> media;
  bool locked;
  bool full;
  String? mediaCreationToken;

  BreedDTO({
    this.id,
    this.currentVersion = '',
    this.kindId = '',
    this.scale = AnimalScale.medium,
    this.investment = 0.0,
    this.territory = 0.0,
    this.pricing = 0.0,
    this.longevity = 0.0,
    this.cohabitation = 0.0,
    this.notes,
    this.title = '',
    this.albumId = '',
    this.locked = true,
    this.full = true,
    this.mediaCreationToken,
    List<ImageDTO>? media,
  }) : media = media ?? [];

  BreedDTO copy() => BreedDTO(
    id: id,
    currentVersion: currentVersion,
    kindId: kindId,
    scale: scale,
    investment: investment,
    territory: territory,
    pricing: pricing,
    longevity: longevity,
    cohabitation: cohabitation,
    notes: notes?.map((n) => n.copy()).toList(),
    title: title,
    albumId: albumId,
    locked: locked,
    full: full,
    mediaCreationToken: mediaCreationToken,
    media: media.map((m) => m.copy()).toList(),
  );

  factory BreedDTO.fromJson(Map<String, dynamic> json) =>
      _$BreedDTOFromJson(json);
  Map<String, dynamic> toJson() => _$BreedDTOToJson(this);
}
